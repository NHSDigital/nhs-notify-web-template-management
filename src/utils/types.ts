export type FormId =
  | 'choose-a-template-type'
  | 'create-nhs-app-template'
  | 'create-email-template'
  | 'create-letter-template'
  | 'create-text-message-template'
  | 'preview-nhs-app-template'
  | 'submit-template'
  | 'create-nhs-app-template-back';

export type FormErrorState = {
  formErrors: string[];
  fieldErrors: Record<string, string[]>;
};

export enum TemplateType {
  NHS_APP = 'NHS_APP',
  SMS = 'SMS',
  EMAIL = 'EMAIL',
  LETTER = 'LETTER',
}

export type Session = {
  id: string;
  templateType: TemplateType | 'UNKNOWN';
  nhsAppTemplateName: string;
  nhsAppTemplateMessage: string;
};

export type FormState = {
  validationError?: FormErrorState;
  redirect?: string;
};

export type TemplateFormState = FormState & Session;

export type PageProps = {
  params: {
    sessionId: string;
  };
};

export type TemplateSubmittedPageProps = {
  params: {
    templateId: string;
  };
};

export type PageComponentProps = {
  initialState: TemplateFormState;
};

export type SubmitTemplatePageComponentProps = {
  templateName: string;
  sessionId: string;
};

export enum TemplateFormatText {
  APP = 'APP',
  EMAIL = 'EMAIL',
  LETTER = 'LETTER',
  SMS = 'SMS',
}
