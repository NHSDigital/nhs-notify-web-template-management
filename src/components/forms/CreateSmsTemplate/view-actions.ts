/**
 * Calculates how many SMS messages are needed to send the given message.
 * This function replicates SMS logic states here
 * https://www.notifications.service.gov.uk/pricing/text-messages#long-text-messages
 * @param {number} characterCount The number of characters in the message.
 * @returns {number} The number of SMS messages needed.
 */
export const calculateHowManySmsMessages = (characterCount: number): number => {
  if (Number.isNaN(characterCount)) return 0;

  if (characterCount > 0 && characterCount <= 160) return 1;

  return Math.ceil(characterCount / 153);
};
