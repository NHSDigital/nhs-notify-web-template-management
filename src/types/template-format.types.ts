export enum TemplateFormatText {
  APP = 'APP',
  EMAIL = 'EMAIL',
  LETTER = 'LETTER',
  SMS = 'SMS',
}
