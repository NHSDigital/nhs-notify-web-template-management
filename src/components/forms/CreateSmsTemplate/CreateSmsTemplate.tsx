'use client';

import { useJsEnabledStyle } from '@hooks/use-js-enabled-style.hook';
import { useTextAreaInput } from '@hooks/use-text-area-input.hook';
import { MessageFormatting } from '@molecules/MessageFormatting/MessageFormatting';
import { NHSNotifyFormWrapper } from '@molecules/NHSNotifyFormWrapper/NHSNotifyFormWrapper';
import { Personalisation } from '@molecules/Personalisation/Personalisation';
import Link from 'next/link';
import {
  Button,
  ChevronLeftIcon,
  HintText,
  Label,
  Textarea,
  TextInput,
} from 'nhsuk-react-components';
import { useFormState } from 'react-dom';
import { PageComponentProps } from '@utils/types';
import { FC } from 'react';
import { ZodErrorSummary } from '@molecules/ZodErrorSummary/ZodErrorSummary';
import { NameYourTemplate } from '@molecules/NameYourTemplate';
import { createSmsTemplateAction } from './server-action';
import { calculateHowManySmsMessages } from './view-actions';

export const CreateSmsTemplate: FC<PageComponentProps> = ({ initialState }) => {
  const [state, action] = useFormState(createSmsTemplateAction, initialState);
  const [smsMessageValue, handler] = useTextAreaInput(
    state.smsTemplateMessage ?? ''
  );
  const templateNameError =
    state.validationError?.fieldErrors.smsTemplateName?.join(', ');

  const templateMessageError =
    state.validationError?.fieldErrors.smsTemplateMessage?.join(', ');

  return (
    <div className='nhsuk-grid-row'>
      <div className='nhsuk-back-link nhsuk-u-margin-bottom-6 nhsuk-u-margin-left-3'>
        <Link
          href={`/choose-a-template-type/${initialState.id}`}
          className='nhsuk-back-link__link'
        >
          <ChevronLeftIcon />
          Go back
        </Link>
      </div>
      <div className='nhsuk-grid-column-two-thirds'>
        <ZodErrorSummary errorHeading='There was a problem' state={state} />
        <h1>Create text message template</h1>
        <NHSNotifyFormWrapper action={action} formId='create-sms-template'>
          <div className={templateNameError && 'nhsuk-form-group--error'}>
            <Label htmlFor='smsTemplateName'>Template name</Label>
            <HintText>This will not be visible to recipients.</HintText>
            <NameYourTemplate template='SMS' />
            <TextInput
              id='smsTemplateName'
              defaultValue={state.smsTemplateName}
              error={templateNameError}
              errorProps={{ id: 'smsTemplateName-error-message' }}
            />
          </div>
          <Textarea
            id='smsTemplateMessage'
            label='Message'
            defaultValue={smsMessageValue}
            onChange={handler}
            maxLength={918}
            rows={10}
            error={templateMessageError}
            errorProps={{ id: 'smsTemplateMessage-error-message' }}
          />
          <div style={useJsEnabledStyle()} id='smsMessageCharacterCount'>
            <p className='nhsuk-u-margin-bottom-0'>
              {smsMessageValue.length} characters
            </p>
            <p>
              This template will be sent as{' '}
              {calculateHowManySmsMessages(Number(smsMessageValue.length))} text
              messages. If you&apos;re using personalisation fields, it could
              send as more.
            </p>
          </div>
          <p>
            <Link
              href='https://notify.nhs.uk/pricing/text-messages'
              target='_blank'
            >
              Learn more about character counts and text messaging pricing
              (opens in a new tab)
            </Link>
          </p>
          <Button id='create-sms-template-submit-button'>Continue</Button>
        </NHSNotifyFormWrapper>
      </div>
      <div className='nhsuk-grid-column-one-third'>
        <Personalisation />
        <MessageFormatting template='SMS' />
      </div>
    </div>
  );
};
