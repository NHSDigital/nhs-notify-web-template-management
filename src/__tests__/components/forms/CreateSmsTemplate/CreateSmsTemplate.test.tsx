import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { mockDeep } from 'jest-mock-extended';
import { TemplateFormState } from '@utils/types';
import { CreateSmsTemplate } from '@forms/CreateSmsTemplate/CreateSmsTemplate';

jest.mock('react-dom', () => {
  const originalModule = jest.requireActual('react-dom');

  return {
    ...originalModule,
    useFormState: (
      _: (
        formState: TemplateFormState,
        formData: FormData
      ) => Promise<TemplateFormState>,
      initialState: TemplateFormState
    ) => [initialState, '/action'],
  };
});

jest.mock('@utils/amplify-utils', () => ({
  getAmplifyBackendClient: () => {},
}));

describe('CreateSmsTemplate component', () => {
  test('renders page', async () => {
    const container = render(
      <CreateSmsTemplate
        initialState={mockDeep<TemplateFormState>({
          validationError: undefined,
          smsTemplateName: '',
          smsTemplateMessage: '',
          id: 'session-id',
        })}
      />
    );
    expect(container.asFragment()).toMatchSnapshot();
  });

  test('renders page with preloaded field values', () => {
    const container = render(
      <CreateSmsTemplate
        initialState={mockDeep<TemplateFormState>({
          validationError: undefined,
          smsTemplateName: 'template-name',
          smsTemplateMessage: 'template-message',
          id: 'session-id',
        })}
      />
    );
    expect(container.asFragment()).toMatchSnapshot();
  });

  test('renders page one error', () => {
    const container = render(
      <CreateSmsTemplate
        initialState={mockDeep<TemplateFormState>({
          validationError: {
            formErrors: [],
            fieldErrors: {
              smsTemplateName: ['Template name error'],
            },
          },
          smsTemplateName: undefined,
          smsTemplateMessage: undefined,
        })}
      />
    );
    expect(container.asFragment()).toMatchSnapshot();
  });

  test('renders page with multiple errors', () => {
    const container = render(
      <CreateSmsTemplate
        initialState={mockDeep<TemplateFormState>({
          validationError: {
            formErrors: [],
            fieldErrors: {
              smsTemplateName: ['Template name error'],
              smsTemplateMessage: ['Template message error'],
            },
          },
          id: 'session-id',
          smsTemplateName: undefined,
          smsTemplateMessage: undefined,
        })}
      />
    );
    expect(container.asFragment()).toMatchSnapshot();
  });

  test('it renders expected character count', async () => {
    const user = userEvent.setup();

    render(
      <CreateSmsTemplate
        initialState={mockDeep<TemplateFormState>({
          validationError: undefined,
          smsTemplateName: '',
          smsTemplateMessage: '',
          id: 'session-id',
        })}
      />
    );

    const templateMessageBox = document.querySelector('#smsTemplateMessage');

    if (!templateMessageBox) {
      throw new Error('Template name box not found');
    }

    const longMessage = 'x'.repeat(300);

    await user.type(templateMessageBox, longMessage);

    const characterCount = document.querySelector('#smsMessageCharacterCount');

    if (!characterCount) {
      throw new Error('Template name box not found');
    }

    expect(characterCount.textContent).toContain(
      `${longMessage.length} characters`
    );
  });
});
