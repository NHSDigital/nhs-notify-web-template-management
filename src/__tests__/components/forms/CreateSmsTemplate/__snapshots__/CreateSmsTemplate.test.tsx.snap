// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateSmsTemplate component renders page 1`] = `
<DocumentFragment>
  <div
    class="nhsuk-grid-row"
  >
    <div
      class="nhsuk-back-link nhsuk-u-margin-bottom-6 nhsuk-u-margin-left-3"
    >
      <a
        class="nhsuk-back-link__link"
        href="/choose-a-template-type/undefined"
      >
        <svg
          aria-hidden="true"
          class="nhsuk-icon nhsuk-icon__chevron-left"
          height="34"
          viewBox="0 0 24 24"
          width="34"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.5 12c0-.3.1-.5.3-.7l5-5c.4-.4 1-.4 1.4 0s.4 1 0 1.4L10.9 12l4.3 4.3c.4.4.4 1 0 1.4s-1 .4-1.4 0l-5-5c-.2-.2-.3-.4-.3-.7z"
          />
        </svg>
        Go back
      </a>
    </div>
    <div
      class="nhsuk-grid-column-two-thirds"
    >
      <h1>
        Create text message template
      </h1>
      <form
        action="/action"
      >
        <input
          name="form-id"
          readonly=""
          type="hidden"
          value="create-sms-template"
        />
        <div>
          <label
            class="nhsuk-label"
            for="smsTemplateName"
          >
            Template name
          </label>
          <div
            class="nhsuk-hint"
          >
            This will not be visible to recipients.
          </div>
          <details
            class="nhsuk-details"
          >
            <summary
              class="nhsuk-details__summary"
            >
              <span
                class="nhsuk-details__summary-text"
              >
                Naming your templates
              </span>
            </summary>
            <div
              class="nhsuk-details__text"
            >
              <p>
                You should name your templates in a way that works best for your service or organisation.
              </p>
              <p>
                Common template names include the:
              </p>
              <ul>
                <li>
                  message channel it uses
                </li>
                <li>
                  subject or reason for the message
                </li>
                <li>
                  intended audience for the template
                </li>
                <li>
                  version number of the template
                </li>
              </ul>
              <p
                data-testid="template-name-example"
              >
                For example, 'SMS - covid19 2023 - over 65s - version 3'
              </p>
            </div>
          </details>
          <div
            class="nhsuk-form-group"
          >
            <input
              class="nhsuk-input"
              id="smsTemplateName"
              name="smsTemplateName"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="nhsuk-form-group"
        >
          <label
            class="nhsuk-label"
            for="smsTemplateMessage"
            id="smsTemplateMessage--label"
          >
            Message
          </label>
          <textarea
            class="nhsuk-textarea"
            id="smsTemplateMessage"
            maxlength="918"
            name="smsTemplateMessage"
            rows="10"
          />
        </div>
        <div
          id="smsMessageCharacterCount"
          style=""
        >
          <p
            class="nhsuk-u-margin-bottom-0"
          >
            0 characters
          </p>
          <p>
            This template will be sent as 0 text messages. If you're using personalisation fields, it could send as more.
          </p>
        </div>
        <p>
          <a
            href="https://notify.nhs.uk/pricing/text-messages"
            target="_blank"
          >
            Learn more about character counts and text messaging pricing (opens in a new tab)
          </a>
        </p>
        <button
          aria-disabled="false"
          class="nhsuk-button"
          id="create-sms-template-submit-button"
          type="submit"
        >
          Continue
        </button>
      </form>
    </div>
    <div
      class="nhsuk-grid-column-one-third"
    >
      <h2
        class="nhsuk-heading-m"
        data-testid="personalisation-header"
      >
        Personalisation
      </h2>
      <details
        class="nhsuk-details"
        data-testid="personalisation-details"
      >
        <summary
          class="nhsuk-details__summary"
          data-testid="personalisation-summary"
        >
          <span
            class="nhsuk-details__summary-text"
          >
            Personalisation fields
          </span>
        </summary>
        <div
          class="nhsuk-details__text"
          data-testid="personalisation-text"
        >
          <p>
            Use double brackets to add a personalisation field to your content. For example:
          </p>
          <code>
            Hello ((firstName)), your NHS number is ((nhsNumber))
          </code>
          <p
            class="nhsuk-u-margin-top-4"
          >
            NHS Notify gets data from PDS to populate personalisation fields.
          </p>
          <p>
            You can use:
          </p>
          <ul>
            <li>
              ((fullName))
            </li>
            <li>
              ((firstName))
            </li>
            <li>
              ((middleNames))
            </li>
            <li>
              ((lastName))
            </li>
            <li>
              ((nhsNumber))
            </li>
            <li>
              ((namePrefix))
            </li>
            <li>
              ((nameSuffix))
            </li>
            <li>
              ((address_line_1))
            </li>
            <li>
              ((address_line_2))
            </li>
            <li>
              ((address_line_3))
            </li>
            <li>
              ((address_line_4))
            </li>
            <li>
              ((address_line_5))
            </li>
            <li>
              ((address_line_6))
            </li>
            <li>
              ((address_line_7))
            </li>
          </ul>
        </div>
      </details>
      <h2
        class="nhsuk-heading-m nhsuk-u-margin-top-4"
        data-testid="personalisation-header"
      >
        Message formatting
      </h2>
      <details
        class="nhsuk-details"
        data-testid="link-and-url-details"
      >
        <summary
          class="nhsuk-details__summary"
          data-testid="link-and-url-summary"
        >
          <span
            class="nhsuk-details__summary-text"
          >
            Links and URLs
          </span>
        </summary>
        <div
          class="nhsuk-details__text"
          data-testid="link-and-url-text"
        >
          <p>
            If the recipient is not expecting to receive a message from you, write the URL in full, starting with https://
          </p>
          <p>
            For example:
          </p>
          <code>
            https://www.nhs.uk/example
          </code>
        </div>
      </details>
    </div>
  </div>
</DocumentFragment>
`;

exports[`CreateSmsTemplate component renders page one error 1`] = `
<DocumentFragment>
  <div
    class="nhsuk-grid-row"
  >
    <div
      class="nhsuk-back-link nhsuk-u-margin-bottom-6 nhsuk-u-margin-left-3"
    >
      <a
        class="nhsuk-back-link__link"
        href="/choose-a-template-type/undefined"
      >
        <svg
          aria-hidden="true"
          class="nhsuk-icon nhsuk-icon__chevron-left"
          height="34"
          viewBox="0 0 24 24"
          width="34"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.5 12c0-.3.1-.5.3-.7l5-5c.4-.4 1-.4 1.4 0s.4 1 0 1.4L10.9 12l4.3 4.3c.4.4.4 1 0 1.4s-1 .4-1.4 0l-5-5c-.2-.2-.3-.4-.3-.7z"
          />
        </svg>
        Go back
      </a>
    </div>
    <div
      class="nhsuk-grid-column-two-thirds"
    >
      <div
        aria-labelledby="error-summary-title"
        class="nhsuk-error-summary"
        role="alert"
        tabindex="-1"
      >
        <h2
          class="nhsuk-error-summary__title"
          data-testid="error-summary"
          id="error-summary-title"
        >
          There is a problem
        </h2>
        <ul
          class="nhsuk-list nhsuk-error-summary__list"
        >
          <li>
            <a
              href="#smsTemplateName"
            >
              Template name error
            </a>
          </li>
          <li>
            <a
              href="#smsTemplateMessage"
            />
          </li>
        </ul>
      </div>
      <h1>
        Create text message template
      </h1>
      <form
        action="/action"
      >
        <input
          name="form-id"
          readonly=""
          type="hidden"
          value="create-sms-template"
        />
        <div
          class="nhsuk-form-group--error"
        >
          <label
            class="nhsuk-label"
            for="smsTemplateName"
          >
            Template name
          </label>
          <div
            class="nhsuk-hint"
          >
            This will not be visible to recipients.
          </div>
          <details
            class="nhsuk-details"
          >
            <summary
              class="nhsuk-details__summary"
            >
              <span
                class="nhsuk-details__summary-text"
              >
                Naming your templates
              </span>
            </summary>
            <div
              class="nhsuk-details__text"
            >
              <p>
                You should name your templates in a way that works best for your service or organisation.
              </p>
              <p>
                Common template names include the:
              </p>
              <ul>
                <li>
                  message channel it uses
                </li>
                <li>
                  subject or reason for the message
                </li>
                <li>
                  intended audience for the template
                </li>
                <li>
                  version number of the template
                </li>
              </ul>
              <p
                data-testid="template-name-example"
              >
                For example, 'SMS - covid19 2023 - over 65s - version 3'
              </p>
            </div>
          </details>
          <div
            class="nhsuk-form-group nhsuk-form-group--error"
          >
            <span
              class="nhsuk-error-message"
              id="smsTemplateName-error-message"
            >
              <span
                class="nhsuk-u-visually-hidden"
              >
                Error: 
              </span>
              Template name error
            </span>
            <input
              aria-describedby="smsTemplateName--error-message"
              class="nhsuk-input nhsuk-input--error"
              id="smsTemplateName"
              name="smsTemplateName"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="nhsuk-form-group"
        >
          <label
            class="nhsuk-label"
            for="smsTemplateMessage"
            id="smsTemplateMessage--label"
          >
            Message
          </label>
          <textarea
            class="nhsuk-textarea"
            id="smsTemplateMessage"
            maxlength="918"
            name="smsTemplateMessage"
            rows="10"
          />
        </div>
        <div
          id="smsMessageCharacterCount"
          style=""
        >
          <p
            class="nhsuk-u-margin-bottom-0"
          >
            0 characters
          </p>
          <p>
            This template will be sent as 0 text messages. If you're using personalisation fields, it could send as more.
          </p>
        </div>
        <p>
          <a
            href="https://notify.nhs.uk/pricing/text-messages"
            target="_blank"
          >
            Learn more about character counts and text messaging pricing (opens in a new tab)
          </a>
        </p>
        <button
          aria-disabled="false"
          class="nhsuk-button"
          id="create-sms-template-submit-button"
          type="submit"
        >
          Continue
        </button>
      </form>
    </div>
    <div
      class="nhsuk-grid-column-one-third"
    >
      <h2
        class="nhsuk-heading-m"
        data-testid="personalisation-header"
      >
        Personalisation
      </h2>
      <details
        class="nhsuk-details"
        data-testid="personalisation-details"
      >
        <summary
          class="nhsuk-details__summary"
          data-testid="personalisation-summary"
        >
          <span
            class="nhsuk-details__summary-text"
          >
            Personalisation fields
          </span>
        </summary>
        <div
          class="nhsuk-details__text"
          data-testid="personalisation-text"
        >
          <p>
            Use double brackets to add a personalisation field to your content. For example:
          </p>
          <code>
            Hello ((firstName)), your NHS number is ((nhsNumber))
          </code>
          <p
            class="nhsuk-u-margin-top-4"
          >
            NHS Notify gets data from PDS to populate personalisation fields.
          </p>
          <p>
            You can use:
          </p>
          <ul>
            <li>
              ((fullName))
            </li>
            <li>
              ((firstName))
            </li>
            <li>
              ((middleNames))
            </li>
            <li>
              ((lastName))
            </li>
            <li>
              ((nhsNumber))
            </li>
            <li>
              ((namePrefix))
            </li>
            <li>
              ((nameSuffix))
            </li>
            <li>
              ((address_line_1))
            </li>
            <li>
              ((address_line_2))
            </li>
            <li>
              ((address_line_3))
            </li>
            <li>
              ((address_line_4))
            </li>
            <li>
              ((address_line_5))
            </li>
            <li>
              ((address_line_6))
            </li>
            <li>
              ((address_line_7))
            </li>
          </ul>
        </div>
      </details>
      <h2
        class="nhsuk-heading-m nhsuk-u-margin-top-4"
        data-testid="personalisation-header"
      >
        Message formatting
      </h2>
      <details
        class="nhsuk-details"
        data-testid="link-and-url-details"
      >
        <summary
          class="nhsuk-details__summary"
          data-testid="link-and-url-summary"
        >
          <span
            class="nhsuk-details__summary-text"
          >
            Links and URLs
          </span>
        </summary>
        <div
          class="nhsuk-details__text"
          data-testid="link-and-url-text"
        >
          <p>
            If the recipient is not expecting to receive a message from you, write the URL in full, starting with https://
          </p>
          <p>
            For example:
          </p>
          <code>
            https://www.nhs.uk/example
          </code>
        </div>
      </details>
    </div>
  </div>
</DocumentFragment>
`;

exports[`CreateSmsTemplate component renders page with multiple errors 1`] = `
<DocumentFragment>
  <div
    class="nhsuk-grid-row"
  >
    <div
      class="nhsuk-back-link nhsuk-u-margin-bottom-6 nhsuk-u-margin-left-3"
    >
      <a
        class="nhsuk-back-link__link"
        href="/choose-a-template-type/undefined"
      >
        <svg
          aria-hidden="true"
          class="nhsuk-icon nhsuk-icon__chevron-left"
          height="34"
          viewBox="0 0 24 24"
          width="34"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.5 12c0-.3.1-.5.3-.7l5-5c.4-.4 1-.4 1.4 0s.4 1 0 1.4L10.9 12l4.3 4.3c.4.4.4 1 0 1.4s-1 .4-1.4 0l-5-5c-.2-.2-.3-.4-.3-.7z"
          />
        </svg>
        Go back
      </a>
    </div>
    <div
      class="nhsuk-grid-column-two-thirds"
    >
      <div
        aria-labelledby="error-summary-title"
        class="nhsuk-error-summary"
        role="alert"
        tabindex="-1"
      >
        <h2
          class="nhsuk-error-summary__title"
          data-testid="error-summary"
          id="error-summary-title"
        >
          There is a problem
        </h2>
        <ul
          class="nhsuk-list nhsuk-error-summary__list"
        >
          <li>
            <a
              href="#smsTemplateName"
            >
              Template name error
            </a>
          </li>
          <li>
            <a
              href="#smsTemplateMessage"
            >
              Template message error
            </a>
          </li>
        </ul>
      </div>
      <h1>
        Create text message template
      </h1>
      <form
        action="/action"
      >
        <input
          name="form-id"
          readonly=""
          type="hidden"
          value="create-sms-template"
        />
        <div
          class="nhsuk-form-group--error"
        >
          <label
            class="nhsuk-label"
            for="smsTemplateName"
          >
            Template name
          </label>
          <div
            class="nhsuk-hint"
          >
            This will not be visible to recipients.
          </div>
          <details
            class="nhsuk-details"
          >
            <summary
              class="nhsuk-details__summary"
            >
              <span
                class="nhsuk-details__summary-text"
              >
                Naming your templates
              </span>
            </summary>
            <div
              class="nhsuk-details__text"
            >
              <p>
                You should name your templates in a way that works best for your service or organisation.
              </p>
              <p>
                Common template names include the:
              </p>
              <ul>
                <li>
                  message channel it uses
                </li>
                <li>
                  subject or reason for the message
                </li>
                <li>
                  intended audience for the template
                </li>
                <li>
                  version number of the template
                </li>
              </ul>
              <p
                data-testid="template-name-example"
              >
                For example, 'SMS - covid19 2023 - over 65s - version 3'
              </p>
            </div>
          </details>
          <div
            class="nhsuk-form-group nhsuk-form-group--error"
          >
            <span
              class="nhsuk-error-message"
              id="smsTemplateName-error-message"
            >
              <span
                class="nhsuk-u-visually-hidden"
              >
                Error: 
              </span>
              Template name error
            </span>
            <input
              aria-describedby="smsTemplateName--error-message"
              class="nhsuk-input nhsuk-input--error"
              id="smsTemplateName"
              name="smsTemplateName"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="nhsuk-form-group nhsuk-form-group--error"
        >
          <label
            class="nhsuk-label"
            for="smsTemplateMessage"
            id="smsTemplateMessage--label"
          >
            Message
          </label>
          <span
            class="nhsuk-error-message"
            id="smsTemplateMessage-error-message"
          >
            <span
              class="nhsuk-u-visually-hidden"
            >
              Error: 
            </span>
            Template message error
          </span>
          <textarea
            aria-describedby="smsTemplateMessage--error-message"
            class="nhsuk-textarea nhsuk-textarea--error"
            id="smsTemplateMessage"
            maxlength="918"
            name="smsTemplateMessage"
            rows="10"
          />
        </div>
        <div
          id="smsMessageCharacterCount"
          style=""
        >
          <p
            class="nhsuk-u-margin-bottom-0"
          >
            0 characters
          </p>
          <p>
            This template will be sent as 0 text messages. If you're using personalisation fields, it could send as more.
          </p>
        </div>
        <p>
          <a
            href="https://notify.nhs.uk/pricing/text-messages"
            target="_blank"
          >
            Learn more about character counts and text messaging pricing (opens in a new tab)
          </a>
        </p>
        <button
          aria-disabled="false"
          class="nhsuk-button"
          id="create-sms-template-submit-button"
          type="submit"
        >
          Continue
        </button>
      </form>
    </div>
    <div
      class="nhsuk-grid-column-one-third"
    >
      <h2
        class="nhsuk-heading-m"
        data-testid="personalisation-header"
      >
        Personalisation
      </h2>
      <details
        class="nhsuk-details"
        data-testid="personalisation-details"
      >
        <summary
          class="nhsuk-details__summary"
          data-testid="personalisation-summary"
        >
          <span
            class="nhsuk-details__summary-text"
          >
            Personalisation fields
          </span>
        </summary>
        <div
          class="nhsuk-details__text"
          data-testid="personalisation-text"
        >
          <p>
            Use double brackets to add a personalisation field to your content. For example:
          </p>
          <code>
            Hello ((firstName)), your NHS number is ((nhsNumber))
          </code>
          <p
            class="nhsuk-u-margin-top-4"
          >
            NHS Notify gets data from PDS to populate personalisation fields.
          </p>
          <p>
            You can use:
          </p>
          <ul>
            <li>
              ((fullName))
            </li>
            <li>
              ((firstName))
            </li>
            <li>
              ((middleNames))
            </li>
            <li>
              ((lastName))
            </li>
            <li>
              ((nhsNumber))
            </li>
            <li>
              ((namePrefix))
            </li>
            <li>
              ((nameSuffix))
            </li>
            <li>
              ((address_line_1))
            </li>
            <li>
              ((address_line_2))
            </li>
            <li>
              ((address_line_3))
            </li>
            <li>
              ((address_line_4))
            </li>
            <li>
              ((address_line_5))
            </li>
            <li>
              ((address_line_6))
            </li>
            <li>
              ((address_line_7))
            </li>
          </ul>
        </div>
      </details>
      <h2
        class="nhsuk-heading-m nhsuk-u-margin-top-4"
        data-testid="personalisation-header"
      >
        Message formatting
      </h2>
      <details
        class="nhsuk-details"
        data-testid="link-and-url-details"
      >
        <summary
          class="nhsuk-details__summary"
          data-testid="link-and-url-summary"
        >
          <span
            class="nhsuk-details__summary-text"
          >
            Links and URLs
          </span>
        </summary>
        <div
          class="nhsuk-details__text"
          data-testid="link-and-url-text"
        >
          <p>
            If the recipient is not expecting to receive a message from you, write the URL in full, starting with https://
          </p>
          <p>
            For example:
          </p>
          <code>
            https://www.nhs.uk/example
          </code>
        </div>
      </details>
    </div>
  </div>
</DocumentFragment>
`;

exports[`CreateSmsTemplate component renders page with preloaded field values 1`] = `
<DocumentFragment>
  <div
    class="nhsuk-grid-row"
  >
    <div
      class="nhsuk-back-link nhsuk-u-margin-bottom-6 nhsuk-u-margin-left-3"
    >
      <a
        class="nhsuk-back-link__link"
        href="/choose-a-template-type/undefined"
      >
        <svg
          aria-hidden="true"
          class="nhsuk-icon nhsuk-icon__chevron-left"
          height="34"
          viewBox="0 0 24 24"
          width="34"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.5 12c0-.3.1-.5.3-.7l5-5c.4-.4 1-.4 1.4 0s.4 1 0 1.4L10.9 12l4.3 4.3c.4.4.4 1 0 1.4s-1 .4-1.4 0l-5-5c-.2-.2-.3-.4-.3-.7z"
          />
        </svg>
        Go back
      </a>
    </div>
    <div
      class="nhsuk-grid-column-two-thirds"
    >
      <h1>
        Create text message template
      </h1>
      <form
        action="/action"
      >
        <input
          name="form-id"
          readonly=""
          type="hidden"
          value="create-sms-template"
        />
        <div>
          <label
            class="nhsuk-label"
            for="smsTemplateName"
          >
            Template name
          </label>
          <div
            class="nhsuk-hint"
          >
            This will not be visible to recipients.
          </div>
          <details
            class="nhsuk-details"
          >
            <summary
              class="nhsuk-details__summary"
            >
              <span
                class="nhsuk-details__summary-text"
              >
                Naming your templates
              </span>
            </summary>
            <div
              class="nhsuk-details__text"
            >
              <p>
                You should name your templates in a way that works best for your service or organisation.
              </p>
              <p>
                Common template names include the:
              </p>
              <ul>
                <li>
                  message channel it uses
                </li>
                <li>
                  subject or reason for the message
                </li>
                <li>
                  intended audience for the template
                </li>
                <li>
                  version number of the template
                </li>
              </ul>
              <p
                data-testid="template-name-example"
              >
                For example, 'SMS - covid19 2023 - over 65s - version 3'
              </p>
            </div>
          </details>
          <div
            class="nhsuk-form-group"
          >
            <input
              class="nhsuk-input"
              id="smsTemplateName"
              name="smsTemplateName"
              type="text"
              value="template-name"
            />
          </div>
        </div>
        <div
          class="nhsuk-form-group"
        >
          <label
            class="nhsuk-label"
            for="smsTemplateMessage"
            id="smsTemplateMessage--label"
          >
            Message
          </label>
          <textarea
            class="nhsuk-textarea"
            id="smsTemplateMessage"
            maxlength="918"
            name="smsTemplateMessage"
            rows="10"
          >
            template-message
          </textarea>
        </div>
        <div
          id="smsMessageCharacterCount"
          style=""
        >
          <p
            class="nhsuk-u-margin-bottom-0"
          >
            16 characters
          </p>
          <p>
            This template will be sent as 1 text messages. If you're using personalisation fields, it could send as more.
          </p>
        </div>
        <p>
          <a
            href="https://notify.nhs.uk/pricing/text-messages"
            target="_blank"
          >
            Learn more about character counts and text messaging pricing (opens in a new tab)
          </a>
        </p>
        <button
          aria-disabled="false"
          class="nhsuk-button"
          id="create-sms-template-submit-button"
          type="submit"
        >
          Continue
        </button>
      </form>
    </div>
    <div
      class="nhsuk-grid-column-one-third"
    >
      <h2
        class="nhsuk-heading-m"
        data-testid="personalisation-header"
      >
        Personalisation
      </h2>
      <details
        class="nhsuk-details"
        data-testid="personalisation-details"
      >
        <summary
          class="nhsuk-details__summary"
          data-testid="personalisation-summary"
        >
          <span
            class="nhsuk-details__summary-text"
          >
            Personalisation fields
          </span>
        </summary>
        <div
          class="nhsuk-details__text"
          data-testid="personalisation-text"
        >
          <p>
            Use double brackets to add a personalisation field to your content. For example:
          </p>
          <code>
            Hello ((firstName)), your NHS number is ((nhsNumber))
          </code>
          <p
            class="nhsuk-u-margin-top-4"
          >
            NHS Notify gets data from PDS to populate personalisation fields.
          </p>
          <p>
            You can use:
          </p>
          <ul>
            <li>
              ((fullName))
            </li>
            <li>
              ((firstName))
            </li>
            <li>
              ((middleNames))
            </li>
            <li>
              ((lastName))
            </li>
            <li>
              ((nhsNumber))
            </li>
            <li>
              ((namePrefix))
            </li>
            <li>
              ((nameSuffix))
            </li>
            <li>
              ((address_line_1))
            </li>
            <li>
              ((address_line_2))
            </li>
            <li>
              ((address_line_3))
            </li>
            <li>
              ((address_line_4))
            </li>
            <li>
              ((address_line_5))
            </li>
            <li>
              ((address_line_6))
            </li>
            <li>
              ((address_line_7))
            </li>
          </ul>
        </div>
      </details>
      <h2
        class="nhsuk-heading-m nhsuk-u-margin-top-4"
        data-testid="personalisation-header"
      >
        Message formatting
      </h2>
      <details
        class="nhsuk-details"
        data-testid="link-and-url-details"
      >
        <summary
          class="nhsuk-details__summary"
          data-testid="link-and-url-summary"
        >
          <span
            class="nhsuk-details__summary-text"
          >
            Links and URLs
          </span>
        </summary>
        <div
          class="nhsuk-details__text"
          data-testid="link-and-url-text"
        >
          <p>
            If the recipient is not expecting to receive a message from you, write the URL in full, starting with https://
          </p>
          <p>
            For example:
          </p>
          <code>
            https://www.nhs.uk/example
          </code>
        </div>
      </details>
    </div>
  </div>
</DocumentFragment>
`;
